lib/SILGen/SILGenDecl.cpp

```diff
@@ -809,50 +809,26 @@ class LetValueInitialization : public Initialization {
                                             SplitCleanups);
   }
 
-  /// This is a helper method for bindValue that handles any changes to the
-  /// value needed for lexical lifetime or no implicit copy purposes.
+  /// This is a helper method for bindValue that creates a scopes operation for
+  /// the lexical variable lifetime and handles any changes to the value needed
+  /// for move-only values.
   SILValue getValueForLexicalLifetimeBinding(SILGenFunction &SGF,
                                              SILLocation PrologueLoc,
                                              SILValue value, bool wasPlusOne) {
-    // If we have none...
-    if (value->getOwnershipKind() == OwnershipKind::None) {
-      // Then check if we have a pure move only type. In that case, we need to
-      // insert a no implicit copy
-      if (value->getType().isMoveOnly(/*orWrapped=*/false)) {
-        value = SGF.B.createMoveValue(PrologueLoc, value, IsLexical);
-        return SGF.B.createMarkUnresolvedNonCopyableValueInst(
-            PrologueLoc, value,
-            MarkUnresolvedNonCopyableValueInst::CheckKind::
-                ConsumableAndAssignable);
-      }
-
-      // If we have a no implicit copy trivial type, wrap it in the move only
-      // wrapper and mark it as needing checking by the move checker.
-      if (vd->isNoImplicitCopy() && value->getType().isTrivial(SGF.F)) {
```

Comment:
@nate-chandler before my rewrite of `getValueForLexicalLifetimeBinding`, noImplicitCopy variables were considered to have lexical lifetimes. Was that intentional? I thought a goal of noImplicitCopy was to be fully optimized, thus eagerly consumed...

Old code:
```
    if (vd->isNoImplicitCopy()) {
      value = SGF.B.createMoveValue(PrologueLoc, value, IsLexical,
                                    DoesNotHavePointerEscape, IsFromVarDecl);
```

Reply:
I don't know.  Looking at the blame, that behavior was added as
```
+      // Otherwise, we have a no implicit copy trivial type, so wrap it in the
+      // move only wrapper and mark it as needing checking by the move cherk.
+      value =
+          SGF.B.createOwnedCopyableToMoveOnlyWrapperValue(PrologueLoc, value);
+      value = SGF.B.createMoveValue(PrologueLoc, value, /*isLexical*/ true);
```
along with the entirety of the original method `getValueForLexicalLifetimeBinding` that seems to have enjoyed mostly NFC changes since then in

```
commit a32c5b3277dbffa6946d117d964e3764d285915e
Date:   Fri Jul 22 11:49:37 2022 -0700

    [silgen] Refactor out a helper method from LetValueInitialization::bindValue.

    Specifically, I refactored out the logic around setting up the value in the case
    of having lexical lifetimes/no implicit copy into the method
    LetValueInitialization::getValueForLexicalLifetimeBinding.

    This logic was taking over bindValue and was making the logic hard to follow.
    Additionally this allows me to take advantage of return value/early exits to
    make the actual logic in getValueForLexicalLifetimeBinding easier to reason
    about.

    NFC.
```

Reply:
> I don't know. Looking at the blame, that behavior was added as

@nate-chandler You don't need to justify the code. I was just wondering if I'm misremembering some design change. As it turns out, yes I was misremembering and I need to revert this change.

Originally no-implicit-copy was going to be eager-consume, but later we changed  course. The only copies that eager-move would be able to eliminate would be explicit copies, and eliminating those is a non-goal. Furthermore, people expect these variables to be destroyed at an explicit `consume` operator, and not earlier than that relative to deinit barriers. To get eager-consume behavior, the programmer simply needs to use an explicit `consume` operator rather that an explicit `copy`.

That's confusing, but basically no-implicit-copy turns on explicit lifetime management. It does *not* turn on automatic aggressive lifetime optimization.